src/onegov/api/token.py

```diff
@@ -13,33 +14,34 @@ def jwt_decode(request: CoreRequest, token: str | bytes) -> Any:
     return jwt.decode(token, request.identity_secret, algorithms=['HS512'])
 
 
-def jwt_encode(request: CoreRequest, payload: dict[str, str]) -> str:
-    iat = utcnow()
+def jwt_encode(request: CoreRequest, payload: dict[str, Any]) -> str:
+
+    iat = datetime.datetime.now(tz=timezone.utc)  # This has to be UTC,
+    # not local
     exp = iat + timedelta(hours=1)
-    headers = {
-        'iat': int(iat.strftime('%s')),
-        'exp': int(exp.strftime('%s')),
-    }
-    return jwt.encode(
-        payload, request.identity_secret, algorithm='HS512', headers=headers
-    )
+    claims = {"iat": iat, "exp": exp}
+    payload.update(claims)
 
+    return jwt.encode(payload, request.identity_secret, algorithm='HS512')
 
-def get_token(request: CoreRequest) -> dict[str, str]:
-    assert request.authorization[0].lower() == 'basic'  # type: ignore[index]
-    auth = b64decode(
-        request.authorization[1].strip()  # type: ignore[index, union-attr]
-    ).decode('utf-8')
 
-    key, _ = auth.split(':', 1)
+def get_token(request: CoreRequest) -> dict[str, str]:
 
-    session = request.session
-    api_key = session.query(ApiKey).filter_by(key=str(key)).one()
+    key = try_get_encoded_token(request)
 
+    api_key = request.session.query(ApiKey).filter_by(key=str(key)).one()
     today = utcnow()
     api_key.last_used = today
-
     payload = {
         'id': str(api_key.id),
     }
     return {'token': jwt_encode(request, payload)}
+
+
+def try_get_encoded_token(request: CoreRequest) -> str:
+    assert request.authorization[0].lower() == 'basic'  # type: ignore[index]
+    auth = b64decode(
+        request.authorization[1].strip()  # type: ignore[index, union-attr]
+    ).decode('utf-8')
```

Comment:
```suggestion
    assert request.authorization is not None
    assert request.authorization.authtype == 'Basic'
    assert isinstance(request.authorization.params, str)
    auth = b64decode(
        request.authorization.params.strip()
    ).decode('utf-8')
```
You can use assertions to perform type narrowing, so you don't need to use type ignores here. Also the authorization is a `NamedTuple`, so you can access the two parts by name.

But this helped me notice that the parsing for the authorization header will only create a dict if `authtype` != `Basic`, so I might be able to replace this with a tagged union so `assert request.authorization.authtype == 'Basic'` will cause the the second item in the tuple to narrow to a `str`. But it might not work until there is `Not` type.

Reply:
Learned about type narrowing today!  Thanks.

Reply:
Another useful tool when messing around with types is the `reveal_type` function, it will cause the type checker to emit a warning with the current type of the variable, so you can observe how the type of a variable changes after you use an assertion to knock out some of the possibilities in a type union.

Reply:
So this is the type that the type checker _believes a variable to be_ at a particular point? That's quite useful.

Reply:
yes